synapse/handlers/saml_handler.py

```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+        try:
+            # Retrieve last session
+            last_session = sessions.pop()
+            # We should have only one session active for one person
+            # This should never match but removing these staled sessions anyway
+            [self._saml_client.local_logout(session["name_id"]) for session in sessions]
+            return last_session
+        except IndexError:
+            return None
+
+    def _find_mxid_from_name_id(self, name_id):
+        try:
+            attributes = self._saml_client.users.get_identity(name_id)
+            for attribute in attributes:
+                if self._mxid_source_attribute in attribute:
+                    return "@%s:%s" % (
+                        attribute[self._mxid_source_attribute][0],
+                        self._auth_handler.hs.hostname,
+                    )
+        except Exception:
+            pass
+        return None
+
+    async def _logout(self, mxid):
+        # first delete all of the user's devices
+        await self._device_handler.delete_all_devices_for_user(mxid)
+
+        # .. and then delete any access tokens which weren't associated with
+        # devices.
+        await self._auth_handler.delete_access_tokens_for_user(mxid)
+
+    # Example: https://github.com/IdentityPython/pysaml2/blob/master/example/sp-wsgi/sp.py
+    def create_logout_request(self, user, access_token):
+        """Create a SAML logout request using HTTP redirect binding
+
+        Returns:
+            bytes: URL to redirect to
+        """
+        logger.info("Creating SAML logout request for %s", user)
+        try:
+            localpart = UserID.from_string(user).localpart
+            logger.debug("User localpart is %s", localpart)
+
+            session = self._find_session_from_user(localpart)
+            # The user probally logged in via m.login.password
+            if session is None:
+                return False
+            logger.debug("User session is %s", session)
+
+            # Creating a logout request through redirect
+            response = self._saml_client.do_logout(
+                session["name_id"],
+                [session["entity"]],
+                reason="/_matrix/client/r0/logout requested",
+                expire=None,
+                expected_binding=BINDING_HTTP_REDIRECT,
+            )
+
+            # Logging out from multiple entities is not supported
+            binding, http_info = next(iter(response.values()))
+            logger.debug("SAML binding %s, http_info %s", binding, http_info)
+
+            redirect_url = next(
+                header[1] for header in http_info["headers"] if header[0] == "Location"
+            )
+            if not redirect_url:
+                raise RuntimeError("missing Location header")
+
+            return redirect_url
+
+        except Exception as e:
+            raise SynapseError(
+                500, "error while creating SAML logout request: %s" % (e,)
+            )
```

Comment:
this is probably redundant?

Reply:
my question is: what is the point of converting it to a SynapseError? if you let the exception propagate, the servlet wrapper will return a 500 error anyway.

On the other hand, if there are "expected" exceptions which shouldn't cause stacktraces, we should probably be returning something other than a 500.